### ISSUE TITLE
📋 Task 2 Created: Add Attention to LSTM

### ISSUE THREAD

**Issue by rootSU5F** [rootSU5F/EduSense-Graduation-Project#20]
### 📋 Description

Currently, the `MultiEmotionOrdinalModel` uses a bidirectional LSTM with simple **global average pooling** to aggregate temporal information. This task is to add an **attention mechanism** that learns to focus on the most important frames for each emotion.

### 🎓 Background

**Current Architecture:**
```
Input (30 frames × 768) → Bi-LSTM → Average Pooling → KAN → 4 Heads
```

**Target Architecture:**
```
Input (30 frames × 768) → Bi-LSTM → Attention Pooling → KAN → 4 Heads
```

**Why Attention?**
- ✅ Learns which frames are important (vs. equal weighting)
- ✅ Different emotions may need focus on different frames
- ✅ Improves interpretability (can visualize attention weights)
- ✅ Better performance on subtle emotions

**Example:**
- **Engagement:** Focus on frames where student looks at screen
- **Confusion:** Focus on frames with furrowed brows
- **Frustration:** Focus on frames with hand gestures

### 📝 Tasks

- [ ] Implement `AttentionPooling` class
- [ ] Modify `MultiEmotionOrdinalModel` to use attention
- [ ] (Optional) Add multi-head attention for each emotion
- [ ] Write unit tests
- [ ] Train for 10 epochs
- [ ] Visualize attention weights
- [ ] Compare results with baseline
- [ ] Update documentation

#### Option 1: Simple Attention (Recommended)

```python
import torch
import torch.nn as nn
import torch.nn.functional as F

class AttentionPooling(nn.Module):
    """
    Attention-based pooling mechanism for sequence data.
    
    Instead of averaging all timesteps equally, learns to weight
    each timestep based on its importance.
    
    Args:
        hidden_dim (int): Dimension of the input features
        
    Input:
        x: (batch, seq_len, hidden_dim)
    Output:
        pooled: (batch, hidden_dim)
        attention_weights: (batch, seq_len) - for visualization
    """
    def __init__(self, hidden_dim):
        super().__init__()
        
        # Learnable attention parameters
        self.attention = nn.Linear(hidden_dim, 1)
    
    def forward(self, x, return_weights=False):
        """
        Args:
            x: Input tensor of shape (batch, seq_len, hidden_dim)
            return_weights: If True, return attention weights for visualization
            
        Returns:
            pooled: Weighted sum of shape (batch, hidden_dim)
            attention_weights: (Optional) Weights of shape (batch, seq_len)
        """
        # Compute attention scores
        # x: (batch, seq_len, hidden_dim)
        # attention_scores: (batch, seq_len, 1)
        attention_scores = self.attention(x)
        
        # Normalize with softmax
        # attention_weights: (batch, seq_len, 1)
        attention_weights = F.softmax(attention_scores, dim=1)
        
        # Weighted sum
        # pooled: (batch, hidden_dim)
        pooled = torch.sum(x * attention_weights, dim=1)
        
        if return_weights:
            return pooled, attention_weights.squeeze(-1)
        return pooled


class MultiEmotionOrdinalModelWithAttention(nn.Module):
    """
    Multi-task ordinal classification with LSTM + Attention.
    
    Args:
        input_dim (int): Input feature dimension (default: 768)
        lstm_hidden (int): LSTM hidden dimension (default: 256)
        lstm_layers (int): Number of LSTM layers (default: 2)
        kan_hidden_dims (list): KAN layer dimensions (default: [128, 64])
        dropout (float): Dropout rate (default: 0.3)
        num_basis (int): Number of basis functions for KAN (default: 8)
        spline_order (int): Spline order for KAN (default: 3)
        num_emotions (int): Number of emotions to predict (default: 4)
    """
    def __init__(self, input_dim=768, lstm_hidden=256, lstm_layers=2,
                 kan_hidden_dims=[128, 64], dropout=0.3,
                 num_basis=8, spline_order=3, num_emotions=4):
        super().__init__()
        
        # LSTM encoder
        self.lstm = nn.LSTM(
            input_size=input_dim,
            hidden_size=lstm_hidden,
            num_layers=lstm_layers,
            batch_first=True,
            dropout=dropout if lstm_layers > 1 else 0,
            bidirectional=True
        )
        
        lstm_output_dim = lstm_hidden * 2  # Bidirectional
        
        # Attention pooling (NEW!)
        self.attention_pooling = AttentionPooling(lstm_output_dim)
        
        # KAN layers
        self.kan_layers = nn.ModuleList()
        current_dim = lstm_output_dim
        for hidden_dim in kan_hidden_dims:
            self.kan_layers.append(KANLayer(current_dim, hidden_dim, num_basis, spline_order))
            self.kan_layers.append(nn.Dropout(dropout))
            current_dim = hidden_dim
        
        # Ordinal heads
        self.ordinal_heads = nn.ModuleList([
            OrdinalHead(current_dim, num_classes=4) for _ in range(num_emotions)
        ])
    
    def forward(self, x, return_attention=False):
        """
        Args:
            x: Input tensor of shape (batch, seq_len, input_dim)
            return_attention: If True, return attention weights for visualization
            
        Returns:
            logits_list: List of 4 tensors, each of shape (batch, 3)
            attention_weights: (Optional) Tensor of shape (batch, seq_len)
        """
        # LSTM encoding
        lstm_out, _ = self.lstm(x)  # (batch, seq_len, lstm_hidden*2)
        
        # Attention pooling (NEW! - replaces mean pooling)
        if return_attention:
            pooled, attention_weights = self.attention_pooling(lstm_out, return_weights=True)
        else:
            pooled = self.attention_pooling(lstm_out)
            attention_weights = None
        
        # KAN layers
        features = pooled
        for layer in self.kan_layers:
            features = layer(features)
        
        # Ordinal heads
        logits_list = [head(features) for head in self.ordinal_heads]
        
        if return_attention:
            return logits_list, attention_weights
        return logits_list


print("✅ Model with attention defined")
```

#### Option 2: Multi-Head Attention (Advanced)

<details>
<summary>Click to expand advanced implementation</summary>

```python
class MultiHeadAttentionPooling(nn.Module):
    """
    Multi-head attention pooling - learns different attention patterns.
    
    Different heads can focus on different aspects:
    - Head 1: Early frames (beginning of video)
    - Head 2: Middle frames (sustained behavior)
    - Head 3: Late frames (final state)
    """
    def __init__(self, hidden_dim, num_heads=4):
        super().__init__()
        assert hidden_dim % num_heads == 0, "hidden_dim must be divisible by num_heads"
        
        self.num_heads = num_heads
        self.head_dim = hidden_dim // num_heads
        
        self.query = nn.Linear(hidden_dim, hidden_dim)
        self.key = nn.Linear(hidden_dim, hidden_dim)
        self.value = nn.Linear(hidden_dim, hidden_dim)
        self.output = nn.Linear(hidden_dim, hidden_dim)
    
    def forward(self, x, return_weights=False):
        batch_size, seq_len, hidden_dim = x.shape
        
        # Project to Q, K, V
        Q = self.query(x).view(batch_size, seq_len, self.num_heads, self.head_dim).transpose(1, 2)
        K = self.key(x).view(batch_size, seq_len, self.num_heads, self.head_dim).transpose(1, 2)
        V = self.value(x).view(batch_size, seq_len, self.num_heads, self.head_dim).transpose(1, 2)
        
        # Compute attention scores
        scores = torch.matmul(Q, K.transpose(-2, -1)) / (self.head_dim ** 0.5)
        attention_weights = F.softmax(scores, dim=-1)
        
        # Apply attention
        attended = torch.matmul(attention_weights, V)
        attended = attended.transpose(1, 2).contiguous().view(batch_size, seq_len, hidden_dim)
        
        # Global pooling over attended sequence
        pooled = attended.mean(dim=1)
        pooled = self.output(pooled)
        
        if return_weights:
            return pooled, attention_weights.mean(dim=1)  # Average across heads
        return pooled
```

</details>

#### Option 3: Emotion-Specific Attention (Most Advanced)

<details>
<summary>Click to expand emotion-specific implementation</summary>

```python
class EmotionSpecificAttention(nn.Module):
    """
    Separate attention mechanism for each emotion.
    
    Rationale: Different emotions focus on different frames.
    - Engagement: Focus on eye contact frames
    - Frustration: Focus on hand gesture frames
    """
    def __init__(self, hidden_dim, num_emotions=4):
        super().__init__()
        
        # Separate attention for each emotion
        self.attention_heads = nn.ModuleList([
            nn.Linear(hidden_dim, 1) for _ in range(num_emotions)
        ])
    
    def forward(self, x, emotion_idx, return_weights=False):
        """
        Args:
            x: (batch, seq_len, hidden_dim)
            emotion_idx: Which emotion to compute attention for (0-3)
        """
        attention_scores = self.attention_heads[emotion_idx](x)
        attention_weights = F.softmax(attention_scores, dim=1)
        pooled = torch.sum(x * attention_weights, dim=1)
        
        if return_weights:
            return pooled, attention_weights.squeeze(-1)
        return pooled
```

</details>

#### Unit Test

```python
def test_attention_model():
    """Test the attention-based model"""
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    
    # Create model
    model = MultiEmotionOrdinalModelWithAttention(
        input_dim=768,
        lstm_hidden=256,
        lstm_layers=2,
        dropout=0.3
    ).to(device)
    
    # Test input
    batch_size = 4
    seq_len = 30
    test_input = torch.randn(batch_size, seq_len, 768).to(device)
    
    # Forward pass without attention weights
    output = model(test_input, return_attention=False)
    assert len(output) == 4
    for out in output:
        assert out.shape == (batch_size, 3)
    
    # Forward pass with attention weights
    output, attention_weights = model(test_input, return_attention=True)
    assert attention_weights.shape == (batch_size, seq_len)
    
    # Check attention weights sum to 1
    attention_sum = attention_weights.sum(dim=1)
    assert torch.allclose(attention_sum, torch.ones(batch_size, device=device), atol=1e-5)
    
    print("✅ All tests passed!")
    print(f"   Attention weights shape: {attention_weights.shape}")
    print(f"   Attention weights sum: {attention_sum[0].item():.6f} (should be 1.0)")


def visualize_attention():
    """Visualize attention weights for a sample"""
    import matplotlib.pyplot as plt
    
    model.eval()
    with torch.no_grad():
        # Get a sample
        sample_idx = 0
        embeddings, labels = val_dataset[sample_idx]
        embeddings = embeddings.unsqueeze(0).to(device)
        
        # Get attention weights
        _, attention_weights = model(embeddings, return_attention=True)
        attention_weights = attention_weights[0].cpu().numpy()
        
        # Plot
        plt.figure(figsize=(12, 4))
        plt.bar(range(len(attention_weights)), attention_weights)
        plt.xlabel('Frame Number', fontsize=12)
        plt.ylabel('Attention Weight', fontsize=12)
        plt.title('Attention Weights Across Frames', fontsize=14, fontweight='bold')
        plt.grid(axis='y', alpha=0.3)
        plt.savefig('attention_weights.png', dpi=300, bbox_inches='tight')
        plt.show()
        
        print(f"✅ Attention visualization saved")
        print(f"   Most attended frame: {attention_weights.argmax()}")
        print(f"   Max attention weight: {attention_weights.max():.4f}")


# Run tests
test_attention_model()
visualize_attention()
```

### 📊 Performance Comparison

After implementation, fill out this comparison table:

| Metric | LSTM (Baseline) | LSTM + Attention | Improvement |
|--------|-----------------|------------------|-------------|
| **Parameters** | ? | ? | ? |
| **Training Time/Epoch** | ? min | ? min | ? |
| **Exact Accuracy** | ? % | ? % | ? |
| **±1 Accuracy** | ? % | ? % | ? |
| **Val Loss** | ? | ? | ? |
| **Engagement Acc** | ? % | ? % | ? |
| **Boredom Acc** | ? % | ? % | ? |
| **Confusion Acc** | ? % | ? % | ? |
| **Frustration Acc** | ? % | ? % | ? |

### 📊 Attention Visualization

After training, analyze which frames the model focuses on:

```python
def analyze_attention_patterns():
    """Analyze which frames are most important for each emotion"""
    
    # Collect attention weights for all validation samples
    all_attention = []
    all_labels = []
    
    model.eval()
    with torch.no_grad():
        for embeddings, labels in val_loader:
            embeddings = embeddings.to(device)
            _, attention = model(embeddings, return_attention=True)
            all_attention.append(attention.cpu())
            all_labels.append(labels)
    
    all_attention = torch.cat(all_attention, dim=0)  # (N, 30)
    all_labels = torch.cat(all_labels, dim=0)  # (N, 4)
    
    # For each emotion level, compute average attention pattern
    emotion_names = ['Engagement', 'Boredom', 'Confusion', 'Frustration']
    
    fig, axes = plt.subplots(2, 2, figsize=(16, 10))
    axes = axes.ravel()
    
    for emotion_idx, emotion_name in enumerate(emotion_names):
        for level in range(4):
            mask = all_labels[:, emotion_idx] == level
            if mask.sum() > 0:
                avg_attention = all_attention[mask].mean(dim=0).numpy()
                axes[emotion_idx].plot(avg_attention, label=f'Level {level}', linewidth=2)
        
        axes[emotion_idx].set_xlabel('Frame Number', fontsize=11)
        axes[emotion_idx].set_ylabel('Average Attention', fontsize=11)
        axes[emotion_idx].set_title(f'{emotion_name} - Attention Patterns', fontsize=13, fontweight='bold')
        axes[emotion_idx].legend()
        axes[emotion_idx].grid(alpha=0.3)
    
    plt.tight_layout()
    plt.savefig('emotion_attention_patterns.png', dpi=300, bbox_inches='tight')
    plt.show()
    
    print("✅ Attention patterns analyzed and saved")
```

### 📚 Resources

- [Effective Approaches to Attention-based Neural Machine Translation](https://arxiv.org/abs/1508.04025)
- [Attention Mechanism Explained](https://towardsdatascience.com/attention-mechanism-in-neural-networks-30aaf5e39512)
- [Visualizing Attention in Transformer Models](https://jalammar.github.io/visualizing-neural-machine-translation-mechanics-of-seq2seq-models-with-attention/)

### ⚠️ Common Issues

<details>
<summary>Click to expand common issues and solutions</summary>

**Issue 1: Attention weights don't sum to 1**
```python
# Problem
attention_weights = torch.sigmoid(attention_scores)  # ❌

# Solution
attention_weights = F.softmax(attention_scores, dim=1)  # ✅
```

**Issue 2: Attention dimension mismatch**
```python
# Make sure to squeeze the last dimension
attention_weights = F.softmax(attention_scores, dim=1)  # (batch, seq, 1)
pooled = torch.sum(x * attention_weights, dim=1)  # ✅ Works due to broadcasting
```

**Issue 3: No improvement over baseline**
- Check if attention weights are actually varying (not uniform)
- Try increasing model capacity (more LSTM layers)
- Use emotion-specific attention (Option 3)

**Issue 4: Attention focuses on only one frame**
- Add regularization: Encourage attention to be more distributed
- Add entropy loss to attention weights

</details>

### 🔍 Debugging Tips

```python
# Check if attention is learning anything useful
def check_attention_diversity():
    """Verify attention weights are not uniform"""
    model.eval()
    with torch.no_grad():
        sample = torch.randn(1, 30, 768).to(device)
        _, attention = model(sample, return_attention=True)
        
        # Compute entropy of attention distribution
        entropy = -(attention * torch.log(attention + 1e-9)).sum()
        
        # Uniform distribution would have entropy = log(30) ≈ 3.4
        print(f"Attention entropy: {entropy.item():.4f}")
        print(f"Uniform entropy: {np.log(30):.4f}")
        
        if entropy > 3.0:
            print("⚠️ Attention might be too uniform (not learning)")
        else:
            print("✅ Attention is focused")
```

### ✅ Acceptance Criteria

- [ ] Code passes all unit tests
- [ ] Model trains without errors for 10 epochs
- [ ] Attention weights sum to 1 (verified)
- [ ] Attention weights visualization generated
- [ ] Results documented in comparison table
- [ ] Improvement over baseline (exact accuracy ≥ baseline)
- [ ] Code follows existing style (type hints, docstrings)

### 🎯 Expected Outcome

- **Better performance** on subtle emotions (Confusion, Boredom)
- **Interpretability** - can see which frames matter
- **Similar training time** (small overhead from attention)
- **Exact accuracy improvement:** +2-5%
- **±1 accuracy improvement:** +1-3%

### 💡 Bonus Tasks (Optional)

- [ ] Implement emotion-specific attention (separate for each emotion)
- [ ] Add attention regularization (entropy penalty)
- [ ] Create interactive attention visualization
- [ ] Compare attention patterns across emotions

### 📅 Timeline

- **Estimated effort:** 3-4 hours
- **Deadline:** [tommorow]
